### ISSUE TITLE
Be able to pass non-scalar tensors through from validation_step to validation_end

### ISSUE THREAD

**Issue by Laksh1997** [PyTorchLightning/pytorch-lightning#716]
## 🚀 Feature
Currently, upon collation of validation steps, there is a line which does `tensor = tensor.item()` **(trainer/logging.py line 90-93)**. This is not good news if the tensor is not a scalar / rank-0 tensor. The solution could be to simply change to `tensor.detach()` if the goal is to just detach the computational graph.

```
        for k, v in callback_metrics.items():
            if isinstance(v, torch.Tensor):
                callback_metrics[k] = v.item()
```

**Comment by williamFalcon** [comment]
@Laksh1997 the  purpose is to  log those keys. whatever you pass back through validation_step should go into validation_end but it has to be a tensor which is a pytorch limitation.

can you give  me an  example where this breaks for  you? will reopen if this needs a code change

**Comment by MarioIshac** [comment]
I think I just ran into the same limitation. I am trying to log a confusion matrix as my metric. I'm interested in logging confusion matrices through train / val / test. This is why I'm implementing it at the metric level. The code fails on the exact line that the OP pointed out, particularly because the tensor is 2D. 

@williamFalcon Does the PyTorch limitation force metrics to be tensors, or scalar tensors? I wonder how big of a change it would be to support non-scalar tensors. Looking through `LightningLoggerBase` the type hints suggest that `log_metrics` only takes `Dict[str, float]`, so looks like the code assumes scalar metrics in atleast couple places.
